src/platform/Platform.c

```diff
@@ -0,0 +1,680 @@
+#include <time.h>
+
+#include <parson.h>
+
+#define MODULE_EXT ".so"
+
+#define AZURE_OSCONFIG "Azure OSConfig"
+
+#define MODEL_VERSION "ModelVersion"
+#define REPORTED "Reported"
+#define COMPONENT_NAME "ComponentName"
+#define OBJECT_NAME "ObjectName"
+
+#define UUID_LENGTH 36
+
+typedef struct MODULE_SESSION
+{
+    MODULE* module;
+    MMI_HANDLE handle;
+
+    struct MODULE_SESSION* next;
+} MODULE_SESSION;
+
+typedef struct SESSION
+{
+    char* uuid;
+    char* client;
+    MODULE_SESSION* modules;
+
+    struct SESSION* next;
+} SESSION;
+
+typedef struct REPORTED_OBJECT
+{
+    const char* component;
+    const char* object;
+} REPORTED_OBJECT;
+
+static SESSION* g_sessions = NULL;
+static MODULE* g_modules = NULL;
+static REPORTED_OBJECT* g_reportedObjects = NULL;
+static int g_numReportedObjects = 0;
+
+void LoadModules(const char* directory, const char* configJson)
+{
+    MODULE* module = NULL;
+    DIR* dir = NULL;
+    struct dirent* entry = NULL;
+    char* path = NULL;
+    char* client = NULL;
+    int version = 0;
+    const char* component = NULL;
+    const char* object = NULL;
+    int reportedCount = 0;
+    JSON_Value* config = NULL;
+    JSON_Object* configObject = NULL;
+    JSON_Array* reportedArray = NULL;
+    JSON_Object* reportedObject = NULL;
+    REPORTED_OBJECT* reported = NULL;
+
+    if (g_modules != NULL)
+    {
+        return;
+    }
+
+    if (NULL == (dir = opendir(directory)))
+    {
+        LOG_ERROR("Failed to open module directory: %s", directory);
+    }
+    else if (NULL == (config = json_parse_file(configJson)))
+    {
+        LOG_ERROR("Failed to parse configuration JSON (%s)", configJson);
+    }
+    else if (NULL == (configObject = json_value_get_object(config)))
+    {
+        LOG_ERROR("Failed to get config object");
+    }
+    else if (0 == (version = json_object_get_number(configObject, MODEL_VERSION)))
+    {
+        LOG_ERROR("Failed to get model version from configuration JSON (%s)", configJson);
+    }
+    else
+    {
+        client = (char*)calloc(strlen(AZURE_OSCONFIG) + strlen(OSCONFIG_VERSION) + 5, sizeof(char));
+        if (NULL == client)
+        {
+            LOG_ERROR("Failed to allocate memory for client name");
+        }
+        else
+        {
+            sprintf(client, "%s %d;%s", AZURE_OSCONFIG, version, OSCONFIG_VERSION);
+        }
+
+        while ((entry = readdir(dir)) != NULL)
+        {
+            if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0)
+            {
+                continue;
+            }
+
+            if (NULL == strstr(entry->d_name, MODULE_EXT))
+            {
+                continue;
+            }
+
+            if (NULL == (path = calloc(strlen(directory) + strlen(entry->d_name) + 2, sizeof(char))))
+            {
+                LOG_ERROR("Failed to allocate memory for module path");
+                continue;
+            }
+
+            sprintf(path, "%s/%s", directory, entry->d_name);
```

Comment:
>entry->d_name [](http://example.com/codeflow?start=46&length=13)

Can entry or d_name be null at this point? Or for the later, empty string? Check